bootstrap/051_materialization_status.sql

```diff
@@ -0,0 +1,158 @@
+
+create table if not exists internal.task_materialization_status(run timestamp_ltz, success boolean, output variant);
+create or replace view reporting.task_materialization_status as select * from internal.task_materialization_status;
+
+create or replace procedure internal.create_materialization_status()
+    returns text
+    language sql
+    execute as owner
+AS
+BEGIN
+    let s text := $$
+    create or replace view admin.materialization_status
+    copy grants
+    as
+    -- The tables we materialize and how often the task refreshes them.
+    with task_tables as (
+        select table_name, user_schema, user_view, task_name from ( values
+            ('QUERY_HISTORY', 'REPORTING', 'ENRICHED_QUERY_HISTORY', 'QUERY_HISTORY_MAINTENANCE'),
+            ('WAREHOUSE_EVENTS_HISTORY', 'REPORTING', 'WAREHOUSE_SESSIONS', 'WAREHOUSE_EVENTS_MAINTENANCE'),
+            ('WAREHOUSE_EVENTS_HISTORY', 'REPORTING', 'CLUSTER_SESSIONS', 'WAREHOUSE_EVENTS_MAINTENANCE'),
+            ('SERVERLESS_TASK_HISTORY', 'REPORTING', 'SERVERLESS_TASK_HISTORY', 'SIMPLE_DATA_EVENTS_MAINTENANCE'),
+            ('TASK_HISTORY', 'REPORTING', 'TASK_HISTORY', 'SIMPLE_DATA_EVENTS_MAINTENANCE'),
+            ('SESSIONS', 'REPORTING', 'SESSIONS', 'SIMPLE_DATA_EVENTS_MAINTENANCE'),
+            ('WAREHOUSE_METERING_HISTORY', 'REPORTING', 'WAREHOUSE_METERING_HISTORY', 'SIMPLE_DATA_EVENTS_MAINTENANCE'),
+            ('LOGIN_HISTORY', 'REPORTING', 'LOGIN_HISTORY', 'SIMPLE_DATA_EVENTS_MAINTENANCE'),
+            ('HYBRID_TABLE_USAGE_HISTORY', 'REPORTING', 'HYBRID_TABLE_USAGE_HISTORY', 'SIMPLE_DATA_EVENTS_MAINTENANCE'),
+            ('MATERIALIZED_VIEW_REFRESH_HISTORY', 'REPORTING', 'MATERIALIZED_VIEW_REFRESH_HISTORY', 'SIMPLE_DATA_EVENTS_MAINTENANCE')
+        ) as t(table_name, user_schema, user_view, task_name)
+    ), task_history as (
+        -- The history rows generated every time a task is run and the table that task is managing
+        select $1 as run, $2 as success, $3 as input, $4 as output, $5 as table_name, $6 as task_name, IFF(INPUT is null, 'FULL', 'INCREMENTAL') as kind FROM (
+            select run::TIMESTAMP_LTZ, success, input, output, 'QUERY_HISTORY', 'QUERY_HISTORY_MAINTENANCE'   from internal.task_query_history
+            union all
+            select run::TIMESTAMP_LTZ, success, input, output, 'WAREHOUSE_EVENTS_HISTORY', 'WAREHOUSE_EVENTS_MAINTENANCE' from internal.task_warehouse_events
+            union all
+            select run::TIMESTAMP_LTZ, success, input, output, table_name, 'SIMPLE_DATA_EVENTS_MAINTENANCE' from internal.task_simple_data_events
+        )
+    ), sf_task_history as (
+        -- Our TASK_HISTORY view is updated with a 3hour delay. Limit how far we back we look in the UDTF.
+        select name as task_name, query_id, graph_run_group_id, state, scheduled_time, QUERY_START_TIME,
+        from table(information_schema.task_history(SCHEDULED_TIME_RANGE_START => TIMESTAMPADD(HOUR, -3, current_timestamp()), result_limit => 10000))
```

Comment:
Having to set `result_limit =10000` isn't great here (and the other UDTF usages) because it's slow and not optimized. However, because we (generally) have a delay of 180minutes for each of the ACCOUNT_USAGE views, we don't have a great solution other than pay this cost.

Because the QH/WEH tasks are running every 1hour, we won't be able to join the query_id/warehouse_size unless we use the UDTF.  At that point, it's just a matter of luck as to how active a user's snowflake account it.

Reply:
> Having to set result_limit =10000 isn't great here (and the other UDTF usages) because it's slow and not optimized. However, because we (generally) have a delay of 180minutes for each of the ACCOUNT_USAGE views, we don't have a great solution other than pay this cost.

The metadata operation for the query_history UDTF with a result_limit of 10K takes 10s on its own. That's a non-starter.

Reply:
Thought: can we push the task_history and query_history UDTF lookups into the respective tasks and emit the query_id and warehouse_size via `output` in the task log tables?